### ISSUE TITLE
Narrowing for constant indexed access fails on array getter

### ISSUE THREAD

**Issue by grant-d** [microsoft/TypeScript#58803]
### 🔎 Search Terms

constant indexed access, array

### 🕗 Version & Regression Information

- This changed between versions `5.4.5` and `5.5.beta`
- This changed in commit or [PR](https://github.com/microsoft/TypeScript/pull/57847)

### ⏯ Playground Link
_I could not repro this in playground, due to the lib import_ 

### 💻 Code

```ts
import crypto from 'crypto-js'

const bar = "abc"
const foo: crypto.lib.WordArray = crypto.enc.Utf8.parse(bar) // Returns type: { words: number[], ... }
for (let i = 0; i < foo.words.length; i++) {
  foo.words[i] += 1
//~~~~~~~~~~~~ error TS2532: Object is possibly 'undefined'.

  // Alternatively:
  foo.words[i] = foo.words[i] + 1
//               ~~~~~~~~~~~~ <same error>
}
```

Adding suitable checks in the loop-body for falsey values does _not_ fix it, eg:
```ts
if (foo && foo.words && foo.words[i]) {
 // Same code as above, same error
}
```

However the following _does_ fix it:
```ts
const word = foo.words[i]
if (word) {
  foo.words[i] = word + 1 // OK
}
```

### 🙁 Actual behavior

error TS2532: Object is possibly 'undefined'.

### 🙂 Expected behavior

No error - this type-checked fine previously

### Additional information about the issue

This issue is new since the new `Control Flow Narrowing for Constant Indexed Accesses` feature. Rolling back to the previous typescript version 'fixes' it

**Comment by Andarist** [comment]
You can often import npm packages in the playground just fine. It's capable of fetching their types. 

I don't see any error here with the nightly version: [TS playground](https://www.typescriptlang.org/play/?ts=5.6.0-dev.20240607#code/JYWwDg9gTgLgBAYygTzDCcBmUIjgciVXQFoArAZ3wChqEIA7C+AIwEMo4BeOAIjZYJedRsywQIALkQo0EAHQAbYC3kB1aABMAglChtk3GcQUBTBgnkBVGJgAc8sBwqmAFOygBKapmhxXiqbwwEYADADccCEAPOIKAO5aFErmAOYwABaRwADUOZ5wAN7UcHHyiVCaFADawAC6cDk8AIy0pQD07XDaijCmUAxsMMAAbqaKyJIlZRVVtQ08vglJ841wrQC+QA). `foo.words` has `number[]` type so I'm not sure how `undefined` has crawled into all of this in your tests. Perhaps you are using `noUncheckedIndexedAccess`?

This would indeed show the difference in the behavior:
[TS 5.4](https://www.typescriptlang.org/play/?noUncheckedIndexedAccess=true&ts=5.4.5#code/PTAEAEDsHsFVIMYAsCmCDWKAmBJSWUAPbAQQQRQGdKAuUAFwCcBXFAKAEsBbAB2kfqgEjAJ4960UADNG0LqADkwsRIC0AK0oK2bBNEiVBAIwCGjUAF5QAIhNGE13fsPTo0OsvHQAdABsORt4A6vxYJIyMJiKWQqJe3iiI3rD0UgAc3jxmlCgAFKaMAJRsUvygub4oghwxAAwA3KA1ADyuPgDuoZR+iQDm9EiNHADUw4WgAN5soG3enYxYlADaHAC6oMNWAIw6MyCgJL70KIyQJvQcAG4oviI007PziyvrVqUdXS8boDsAvkA)
[TS nightly](https://www.typescriptlang.org/play/?noUncheckedIndexedAccess=true&ts=5.6.0-dev.20240607&ssl=1&ssc=1&pln=1&pc=35#code/PTAEAEDsHsFVIMYAsCmCDWKAmBJSWUAPbAQQQRQGdKAuUAFwCcBXFAKAEsBbAB2kfqgEjAJ4960UADNG0LqADkwsRIC0AK0oK2bBNEiVBAIwCGjUAF5QAIhNGE13fsPTo0OsvHQAdABsORt4A6vxYJIyMJiKWQqJe3iiI3rD0UgAc3jxmlCgAFKaMAJRsUvygub4oghwxAAwA3KA1ADyuPgDuoZR+iQDm9EiNHADUw4WgAN5soG3enYxYlADaHAC6oMNWAIw6MyCgJL70KIyQJvQcAG4oviI007PziyvrVqUdXS8boDsAvkA)

This, in turn, looks like a correctness improvement.

**Comment by RyanCavanaugh** [comment]
If this is just an array, I don't see why you need `crypto-js` as part of the repro. Please provide a minimal example, with config, that shows what's happening

**Comment by grant-d** [comment]
@RyanCavanaugh yes I am aware of that, but I was unable to find a repro without the package.

@Andarist I think you nailed it - thanks so much, that must be the issue. Not sure what the action is here, I am closing but please re-open as necessary.

**Comment by grant-d** [comment]
@RyanCavanaugh here is a minimal repro in [TS Playground](https://www.typescriptlang.org/play/?noUncheckedIndexedAccess=true&ssl=9&ssc=2&pln=1&pc=1#code/PTAEAEDsHsFVIMYAsCmCDWKAmBJSWUAPbAQQQRQGdKAuUAFwCcBXFAKAEtJ6VGAzAIYVQAMWjRQAb1AB3aIyy1QkZgFsARrwDaAXVABfNgmiRK9UH3F0xEgLxTZ8xXS0BGADSgATJ4DMewzZLRlAACgAbFHMOUHsABgBuUBiAHgtxADo5BUoMyMgAc3okJI4AajKASik2UGS+MMtoLKdKLQ4daslauvTm7MV2vXsmlpyh0DLQVx7DQyA) (based on @Andarist repro)

Including code here for ease of understanding:
```typescript
// @noUncheckedIndexedAccess: true
interface Foo { words: number[] }
const foo: Foo = { words: [1, 2, 3] }

for (let i = 0; i < foo.words.length; i++) {
  if (foo.words[i]) {
    foo.words[i] = foo.words[i] + 1
    //             ^Object is possibly 'undefined'.(2532)^
  }
}
```
<img width="530" alt="image" src="https://github.com/microsoft/TypeScript/assets/10214147/200a9116-aa66-4b77-9712-d95eb1bb6f14">

**Comment by Andarist** [comment]
The problem here is that `i` is assigned to so it doesn't pass this check: 
https://github.com/microsoft/TypeScript/pull/57847/files#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8R26613

`i` is special in loops. We know that it's only assigned to after this loop block. So perhaps some extra fine-tuning for this case would be possible.

**Comment by Andarist** [comment]
Oh, this just never works for variables that are reassigned:
```ts
function f1(obj: Record<string, unknown>) {
  let key: string = "";
  if (typeof obj[key] === "string") {
    obj[key].toUpperCase();
  }
  key = ""; // toggle to see the error to go away and come back
}
```

This is kinda spooky to see an error pop up on a previous line that seemingly has no real correlation with the line that just got added.

**Comment by grant-d** [comment]
Possibly related 
https://github.com/microsoft/TypeScript/issues/59185
